### PR TITLE
batches: add and parse site configuration for rollout windows

### PR DESCRIPTION
**Pull request by LawnGnome**
Part of #18921.

This PR just adds the bits we need to parse and validate the rollout window configuration. Using that configuration to construct a scheduler (and, you know, writing a scheduler) is coming in future PRs.

### CODE REVIEW

File: enterprise/internal/batches/config.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package batches
+
+import (
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/batches/scheduler/window"
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+)
+
+func init() {
+	conf.ContributeValidator(func(c conf.Unified) (problems conf.Problems) {
```
Comment (mrnugget) [inline enterprise/internal/batches/config.go:9]: Not a blocker, but I'm wondering: can we move this to [`InitFrontend`](https://github.com/sourcegraph/sourcegraph/blob/5d985255e0fd6be1daec6f2cce05a31ee14a850b/enterprise/internal/batches/frontend.go#L19)? (Asking because, if possible, I'd love to avoid `init()` functions)

File: enterprise/internal/batches/scheduler/window/config.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+}
+
+// ValidateConfiguration validates the given site configuration.
+func ValidateConfiguration(raw *[]*schema.BatchChangeRolloutWindow) error {
+	return (&Configuration{
+		windows: []Window{},
+	}).updateFromConfig(raw)
+}
+
+func (cfg *Configuration) updateFromConfig(raw *[]*schema.BatchChangeRolloutWindow) error {
```
Comment (mrnugget) [inline enterprise/internal/batches/scheduler/window/config.go:38]: Nitpick, but `cfg.updateFromConfig` is a bit weird 😄 

```suggestion
func (cfg *Configuration) updateFromRaw(raw *[]*schema.BatchChangeRolloutWindow) error {
```

File: enterprise/internal/batches/scheduler/window/config.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+// NewConfiguration constructs a Configuration based on the current site
+// configuration, including watching the configuration for updates.
+func NewConfiguration() *Configuration {
+	cfg := &Configuration{}
+
+	conf.Watch(func() {
+		if err := cfg.updateFromConfig(conf.Get().BatchChangesRolloutWindows); err != nil {
+			log15.Warn("ignoring erroneous batchChanges.rolloutWindows configuration", "err", err)
+		}
```
Comment (mrnugget) [inline enterprise/internal/batches/scheduler/window/config.go:25]: Do you need a mutex around `windows`? If this goroutine fires in the background while someone's holding a `*Configuration` then there's a race condition.

**Review by mrnugget — APPROVED** [review]
Good stuff! Especially like the tests and how the scope of the PR 👍

File: enterprise/internal/batches/scheduler/window/config.go [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+)
+
+// Configuration represents the rollout windows configured on the site.
+type Configuration struct {
+	windows []Window
+}
+
+// NewConfiguration constructs a Configuration based on the current site
+// configuration, including watching the configuration for updates.
+func NewConfiguration() *Configuration {
```
Comment (eseliger) [inline enterprise/internal/batches/scheduler/window/config.go:19]: Is it weird when you use the configuration in, say, a loop and it changes while looping over it? I would imagine it needs to be "locked" before using it and "unlocked" after. Since that can be forgotten easily, is it maybe simpler to just load the configuration from the config once `NewConfiguration` is called and the call site needs to get a new configuration, when they desire to "reload"?

**Review by eseliger — APPROVED** [review]
LGTM, just a question around the auto-reload thingy

File: enterprise/internal/batches/scheduler/window/config.go (hunk 4 above)
Comment (mrnugget) [reply enterprise/internal/batches/scheduler/window/config.go:19]: > Since that can be forgotten easily, is it maybe simpler to just load the configuration from the config once `NewConfiguration` is called and the call site needs to get a new configuration, when they desire to "reload"?

I like that!

File: enterprise/internal/batches/scheduler/window/config_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			want int
+		}{
+			"one bad window": {
+				in: &[]*schema.BatchChangeRolloutWindow{
+					{Rate: "xx"},
+					{Rate: 0},
+				},
+				want: 1,
+			},
+			"two bad windows, ha ha ha": {
```
Comment (chrispine) [inline enterprise/internal/batches/scheduler/window/config_test.go:31]: ha ha ha!

**Review by chrispine — APPROVED** [review]
LGTM

File: enterprise/internal/batches/config.go (hunk 1 above)
Comment (LawnGnome) [reply enterprise/internal/batches/config.go:9]: I did it this way [to align with other parts of the codebase that validate configuration at runtime](https://sourcegraph.com/search?q=ContributeValidator+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+&patternType=literal). I'm not particularly against making the move, but it doesn't feel aligned with what `InitFrontend` does today.

How strongly do you feel about this? Should we be migrating other cases like this too?

File: enterprise/internal/batches/scheduler/window/config.go (hunk 2 above)
Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:38]: Yeah, you can really see the seams there where I did a last minute rename of a type. I'll change it to just `update`. :smile:

File: enterprise/internal/batches/scheduler/window/config.go (hunk 3 above)
Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:25]: Yes. Good catch, thanks!

Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:25]: Actually, I'm revising my "yes" to "no", but it still needs a change: we can make the change in `update` atomic. (More on concurrency below in response to Erik's review.)

File: enterprise/internal/batches/scheduler/window/config.go (hunk 4 above)
Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:19]: I'm not so keen: the string parsing needed to build a `Configuration` is simple and cheap, but not free, and I don't expect window configuration to change often in practice, so it doesn't feel like something I'd want us to be doing every time the scheduler wakes up or a user loads a batch change page.

I think the issue here is that I haven't made it clear in this PR how this is going to get used by higher level code: I don't expect the scheduler to implement locking around the configuration — I agree that would be a poor choice — but I (possibly foolishly) think it's possible to expose a safe API from `Configuration` that can still handle live updates.

So, in summary, let's table this discussion (and PR) for the moment, and let's see if I can make it clearer once there's more context available soon.

Comment (eseliger) [reply enterprise/internal/batches/scheduler/window/config.go:19]: Just as a note: I'm not against merging this PR btw, it's totally fine as-is and if it unblocks you, go for it. I just wanted to mention the thought because it struck my mind when I read it, but obviously I don't have as much insight into whats to come than you have!

**Comment by LawnGnome** [comment]
No; I wanted to split this project up in a fairly fine-grained manner, but
thinking about this in response to your and Thorsten's feedback made me
realise I couldn't articulate this well enough yet, so I think it's
incumbent on me to do a bit more work on this before we lock too much in.

File: enterprise/internal/batches/scheduler/window/config.go (hunk 4 above)
Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:19]: No; I wanted to split this project up in a fairly fine-grained manner, but thinking about this in response to your and Thorsten's feedback made me realise I couldn't articulate this well enough yet, so I think it's incumbent on me to do a bit more work on this before we lock too much in.

File: enterprise/internal/batches/scheduler/window/config.go (hunk 3 above)
Comment (LawnGnome) [reply enterprise/internal/batches/scheduler/window/config.go:25]: The final update: OK, we do need a mutex, because [variable assignment in Go is never atomic](https://stackoverflow.com/questions/21447463/is-assigning-a-pointer-atomic-in-go). :grimacing:

File: enterprise/internal/batches/config.go (hunk 1 above)
Comment (mrnugget) [reply enterprise/internal/batches/config.go:9]: Not super strongly - it's not a blocker. But I do think it would be good to do. The other use cases are also older than this one and most likely don't have an `InitFrontend` function available, which is also why those can't be easily converted to another pattern. It's just that we have this function and it's our entry point to frontend/global state, so it would be good to have this in there.

**Comment by LawnGnome** [comment]
I was going to force-push this and reopen it for review, but I think it's now changed enough that it's really a new PR, so I'm going to close this.
